plugins/catalog-backend-module-github/src/providers/GithubEntityProvider.ts

```diff
@@ -0,0 +1,313 @@
+ *
+ * @public
+ */
+export class GithubEntityProvider implements EntityProvider {
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+  private readonly githubCredentialsProvider: GithubCredentialsProvider;
+
+  static fromConfig(config: Config, options: GithubEntityProviderOptions) {
+    const integrations = ScmIntegrations.fromConfig(config);
+    if (!integrations) {
+      throw new Error('No integrations found for github');
+    }
+    return new GithubEntityProvider(integrations, options);
+  }
+
+  private constructor(
+    private integrations: ScmIntegrations,
+    private options: {
+      id: string;
+      orgUrl: string;
+      files: Array<string>;
+      logger: Logger;
+      schedule: TaskRunner;
+      githubCredentialsProvider?: any;
+    },
+  ) {
+    options.logger.child({ target: this.getProviderName() });
+
+    this.githubCredentialsProvider =
+      options.githubCredentialsProvider ||
+      DefaultGithubCredentialsProvider.fromIntegrations(integrations);
+
+    this.scheduleFn = this.createScheduleFn(options.schedule);
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `github-entity-provider:${this.options.id}`;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    return await this.scheduleFn();
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.options.logger.child({
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const gitHubConfig = this.integrations.github.byUrl(
+      this.options.orgUrl,
+    )?.config;
+
+    if (!gitHubConfig) {
+      throw new Error(
+        `There is no GitHub integration that matches ${this.options.orgUrl}. Please add a configuration entry for it under integrations.github`,
+      );
+    }
+
+    const { org, repoSearchPath } = parseUrl(this.options.orgUrl);
+
+    const { headers } = await this.githubCredentialsProvider.getCredentials({
+      url: this.options.orgUrl,
+    });
+
+    const client = graphql.defaults({
+      baseUrl: gitHubConfig.apiBaseUrl,
+      headers,
+    });
+
+    const startTimestamp = Date.now();
+
+    logger.info(`Reading GitHub repositories from ${this.options.orgUrl}`);
+
+    if (!this.options.files) {
```

Comment:
this should also have been checked upfront immediately in fromConfig, why allow broken inputs that then fail at runtime :) same thing with checking that the array isn't empty

Reply:
resolved